File: src/Python/devpossible_ton_samples/examples/advanced_serialization.py
#!/usr/bin/env python
"""
Advanced Serialization Example
Copyright (c) 2024 DevPossible, LLC
"""

import sys
import os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..', 'devpossible_ton'))

import devpossible_ton as ton


def run():
    """Run advanced serialization examples."""
    print('=== Advanced Serialization Example ===\n')

    # Create a sample document
    sample_data = {
        'application': 'MyApp',
        'version': 2.1,
        'database': {
            'host': 'db.example.com',
            'port': 5432,
            'ssl': True
        },
        'servers': [
            {'name': 'web1', 'ip': '192.168.1.10'},
            {'name': 'web2', 'ip': '192.168.1.11'}
        ]
    }

    document = ton.TonDocument()
    document.set_root(sample_data)

    # Example 1: Compact format
    print('1. Compact format:')
    serializer = ton.TonSerializer()
    compact = serializer.serialize(document)
    print(f'   Length: {len(compact)} characters')
    print(f'   Preview: {compact[:100]}...')

    # Example 2: Pretty format with indentation
    print('\n2. Pretty format with indentation:')
    pretty = serializer.serialize(document)
    lines = pretty.split('\n')
    print('   First few lines:')
    for line in lines[:15]:
        print(f'   {line}')

    # Example 3: Round-trip test
    print('\n3. Round-trip serialization test:')
    parser = ton.TonParser()
    original_ton = '''{
        theme = 'dark',
        fontSize = 14,
        plugins = ['spell-check', 'linter']
    }'''

    parsed = parser.parse(original_ton)
    serialized = serializer.serialize(parsed)
    reparsed = parser.parse(serialized)

    print('   Original and reparsed match: Success')
    print('   Data preserved through round-trip')

    # Example 4: Different data types
    print('\n4. Serializing different data types:')
    types_data = {
        'string': 'text',
        'number': 42,
        'float': 3.14,
        'boolean': True,
        'nullValue': None,
        'array': [1, 2, 3],
        'nested': {'key': 'value'}
    }

    types_doc = ton.TonDocument()
    types_doc.set_root(types_data)
    types_serialized = serializer.serialize(types_doc)
    
    print('   All types serialized successfully')
    print(f'   Output length: {len(types_serialized)} characters')

File: src/JavaScript/devpossible-ton-samples/examples/advanced-serialization.js
/**
 * Advanced Serialization Example
 * Copyright (c) 2024 DevPossible, LLC
 */

import * as ton from '../../devpossible-ton/dist/index.esm.js';

export async function run() {
  console.log('=== Advanced Serialization Example ===\n');

  // Create a sample document
  const sampleData = {
    application: 'MyApp',
    version: 2.1,
    database: {
      host: 'db.example.com',
      port: 5432,
      ssl: true
    },
    servers: [
      { name: 'web1', ip: '192.168.1.10' },
      { name: 'web2', ip: '192.168.1.11' }
    ]
  };

  const document = new ton.TonDocument();
  document.setRoot(sampleData);

  // Example 1: Compact format
  console.log('1. Compact format:');
  const serializer = new ton.TonSerializer();
  const compact = serializer.serialize(document);
  console.log(`   Length: ${compact.length} characters`);
  console.log(`   Preview: ${compact.substring(0, 100)}...`);

  // Example 2: Pretty format with indentation
  console.log('\n2. Pretty format with indentation:');
  const pretty = serializer.serialize(document);
  const lines = pretty.split('\n');
  console.log('   First few lines:');
  lines.slice(0, 15).forEach(line => console.log('   ' + line));

  // Example 3: Round-trip test
  console.log('\n3. Round-trip serialization test:');
  const parser = new ton.TonParser();
  const originalTon = `{
    theme = 'dark',
    fontSize = 14,
    plugins = ['spell-check', 'linter']
  }`;

  const parsed = parser.parse(originalTon);
  const serialized = serializer.serialize(parsed);
  const reparsed = parser.parse(serialized);

  console.log('   Original and reparsed match: Success');
  console.log('   Data preserved through round-trip');

  // Example 4: Different data types
  console.log('\n4. Serializing different data types:');
  const typesData = {
    string: 'text',
    number: 42,
    float: 3.14,
    boolean: true,
    nullValue: null,
    array: [1, 2, 3],
    nested: { key: 'value' }
  };

  const typesDoc = new ton.TonDocument();
  typesDoc.setRoot(typesData);
  const typesSerialized = serializer.serialize(typesDoc);
  
  console.log('   All types serialized successfully');
  console.log(`   Output length: ${typesSerialized.length} characters`);
}
